### ISSUE TITLE
Making it work on Android

### ISSUE THREAD

**Issue by TianZerL** [bloc97/Anime4K#99]
Hi, I just found that the Anime4K glsl shaders can't run in mpv-android, it will be a blue screen.  

After some testing, I found that it should be because of the max and min opengl built-in functions.  

In the mpv-android, code like `max(a, 0) //a is a vec4` is unable to compile successfully, it should be `max(a, vec4(0))` or `max(a, vec4(0.0f))`. It works for me.  

Although the shader works, it doesn't feel as good as PC.
Any ideas?

**Comment by yushengnan1** [comment]
@TianZerL Hi, I just installed mpv-android,but I have some problem in the installation of Anime4k glsl. What is the installation path of glsl file in mpv-android? Or install it to any location in the memory and call the path directly?Could you please help me ?I can pay for this. Thank you!

**Comment by TianZerL** [comment]
@yushengnan1
Just unzip it to any location, and edit the config file, it is same as using it on a PC.

**Comment by Cyberspiri** [comment]
@yushengnan1 Hi! so did you managed to get it work?

**Comment by GizmoTheGreen** [comment]
I can't figure out where to add the glsl files on android. or how to set up the mpv.conf to find them. any example would be helpful.

**Comment by bloc97** [comment]
@GizmoTheGreen 
You can add the glsl files anywhere if I am not mistaken. You would need to write absolute paths, eg. `/home/user/shaders/Anime4K_Upscale_CNN_x2_S.glsl`.

mpv.conf can be edited with a file manager or by going in mpv settings -> advanced -> edit mpv.conf or edit input.conf

**Comment by bloc97** [comment]
To find the absolute path of a folder you can go to the files app.
<img src="https://user-images.githubusercontent.com/567732/129249102-b0d1dfe7-41c9-4dab-bd0a-3071941eaae9.png" width="200" />

<img src="https://user-images.githubusercontent.com/567732/129249298-0dead014-345f-48bc-a0d3-2499f1652522.png" width="200" />

If I were to put all the shaders under a folder named `shaders` in `Documents`, I would write in mpv.conf:
`glsl-shaders="/storage/emulated/0/shaders/Anime4K_Upscale_CNN_x2_S.glsl"`

**Comment by FalaaY** [comment]
Hi. Trying to use this on My nvidia shield Pro 2019.

Could someone who has successfully done it on android just sent over the configuration file they used then I can edit, driving me insane trying to get it working.

**Comment by arianaa30** [comment]
Can someone help me with my Android setup as well? I edited the `mpv.conf` with paths, but how to modify `input.conf`? In Windows setup, it uses ctrl + 1 and such which we don't have in Android. What exactly should I enter there?

**Comment by sami155** [comment]
> Can someone help me with my Android setup as well? I edited the `mpv.conf` with paths, but how to modify `input.conf` and enable the shader? In Windows setup, it uses ctrl + 1 and such which we don't have in Android. What exactly should I enter there?

Having the same issue

**Comment by yooooooLeon** [comment]
您的邮件已收到！谢谢！

**Comment by sami155** [comment]
I finally got it omg. The translations from the aniyomi settings made me understand a few things wrong. Some docs and this thread made it more complicated than it is.
Will post the explanation tomorrow im going to sleep

**Comment by sami155** [comment]
Its easier than thought.
# For Aniyomi:

1. Get into the settings and to Player. Then to advanced player settings. 
~~Then enable "Enable MPV Scripts"~~
Apparently you dont need to activate MPV Scripts.
2. At the very top you will find "edit mpv configuration file for further player settings"
3. There you need to insert the mpv script for your favorite mode. I have prepared/converted Mode A, B and C down below.

//Mode A
```
# Optimized shaders for lower-end GPU: Mode A (Fast)
profile=gpu-hq
glsl-shaders="/storage/emulated/0/Documents/shaders/Anime4K_Clamp_Highlights.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Restore_CNN_M.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_M.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x2.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x4.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_S.glsl"
```

//Mode B
```
# Optimized shaders for lower-end GPU: Mode B (Fast)
profile=gpu-hq
glsl-shaders="/storage/emulated/0/Documents/shaders/Anime4K_Clamp_Highlights.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Restore_CNN_Soft_M.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_M.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x2.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x4.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_S.glsl"
```

//Mode C
```
# Optimized shaders for lower-end GPU: Mode C (Fast)
profile=gpu-hq
glsl-shaders="/storage/emulated/0/Documents/shaders/Anime4K_Clamp_Highlights.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_Denoise_CNN_x2_M.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x2.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x4.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_S.glsl"
```


Beware that the GPU Load gets to 85-100% instead of 5-15%. 
My phone (Samsung S22) as an example, gets hot very fast for some Animes. And the battery also drains faster.
Just keep that in mind.


# For MPV Android

I couldn't find a way 🥲

**Comment by rowel-eshan** [comment]
> # For Aniyomi:



Just did as you said. Just wondering how I'd go about enabling the shader. Should I add smth in the keybinds section?

**Comment by sami155** [comment]
> > # For Aniyomi:
> 
> 
> 
> Just did as you said. Just wondering how I'd go about enabling the shader. Should I add smth in the keybinds section?

No. It should work without any additional steps. Just play a video and show the second statistics page. (3 dots in the player)

Then you should get every shader listed

**Comment by rowel-eshan** [comment]
Here is what I see when I show the 2nd page of stats...

![Image](https://github.com/user-attachments/assets/bd86bf92-91fe-4d5c-a8b8-f55198016aa5)

![Image](https://github.com/user-attachments/assets/a030029a-c202-4471-85e0-3323d24bcd94)

![Image](https://github.com/user-attachments/assets/ef47c264-7df3-45ef-b463-1d827bce2bb9)

My file structure for the script locations match the config too. I tried pasting in each mode. Doesn't seem to work... Not sure what I'm doing wrong :(


Using an s24 ultra btw

**Comment by sami155** [comment]
@rowel-eshan ok, maybe its the formatting? can you please try this?

```
# Optimized shaders for lower-end GPU: Mode B (Fast)
profile=gpu-hq
glsl-shaders="/storage/emulated/0/Documents/shaders/Anime4K_Clamp_Highlights.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Restore_CNN_Soft_M.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_M.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x2.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x4.glsl"
glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_S.glsl"
```

**Comment by rowel-eshan** [comment]
@sami155
 
Holy shit it works! It absolutely was the formatting. Thank you so much!!!!

Note: opening this page in mobile view was inserting some newlines I believe.

**Comment by sami155** [comment]
@rowel-eshan thanks for the fast update. will update my solution to code blocks
edit: done

**Comment by rowel-eshan** [comment]
> [@rowel-eshan](https://github.com/rowel-eshan) thanks for the fast update. will update my solution to code blocks edit: done

Thanks! I have Anime4k working on my pc but for years I always wished I could watch it on my phone! Now I definitely can 🙏

**Comment by aidan-walden** [comment]
After a lot of troubleshooting I've been able to get it to work using mpv-android. You need to use ADB in order to set everything up correctly. Here's how:

- Connect to your device using ADB
- Create the directory `/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders` using mkdir
- Place all of the shaders in this directory (I used mv to move them from the downloads folder)
- In mpv, edit mpv.conf to use the shaders, using absoulte paths (so `~~/shaders` becomes `/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders`)

Here is my mpv.conf for the low-end shaders (this will automatically apply the shaders upon playback):
`
glsl-shaders="/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders/Anime4K_Clamp_Highlights.glsl:/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders/Anime4K_Restore_CNN_M.glsl:/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders/Anime4K_Upscale_CNN_x2_M.glsl:/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders/Anime4K_AutoDownscalePre_x2.glsl:/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders/Anime4K_AutoDownscalePre_x4.glsl:/storage/emulated/0/Android/data/is.xyz.mpv/files/shaders/Anime4K_Upscale_CNN_x2_S.glsl"`

**Comment by akeakeakeake** [comment]
> [@rowel-eshan](https://github.com/rowel-eshan) ok, maybe its the formatting? can you please try this?
> 
> ```
> # Optimized shaders for lower-end GPU: Mode B (Fast)
> profile=gpu-hq
> glsl-shaders="/storage/emulated/0/Documents/shaders/Anime4K_Clamp_Highlights.glsl"
> glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Restore_CNN_Soft_M.glsl"
> glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_M.glsl"
> glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x2.glsl"
> glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x4.glsl"
> glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_S.glsl"
> ```

Do you have any idea why this doesn't work for nvidia shield? Should be same path unless i'm missing something

![Image](https://github.com/user-attachments/assets/5afbbaab-1b92-41ec-be94-4c2922dda5b1)

![Image](https://github.com/user-attachments/assets/988b170a-ec55-4765-b0dc-742453b5ce35)

**Comment by akeakeakeake** [comment]
My current mpv.conf after trying like 20 different types:

![Image](https://github.com/user-attachments/assets/f7e2334f-a828-4e54-bd51-395be324a481)

My input.conf (Connecting a keyboard and pressing 1/3 to enable/disable the shaders also does nothing apart from showing the text):

![Image](https://github.com/user-attachments/assets/d8bf5cee-a379-4b64-923c-027b152194aa)

**Comment by sami155** [comment]
> My current mpv.conf after trying like 20 different types:
> 
> ![Image](https://github.com/user-attachments/assets/f7e2334f-a828-4e54-bd51-395be324a481)
> 
> My input.conf (Connecting a keyboard and pressing 1/3 to enable/disable the shaders also does nothing apart from showing the text):
> 
> ![Image](https://github.com/user-attachments/assets/d8bf5cee-a379-4b64-923c-027b152194aa)

Did you try not using input.config and leave it empty?
Edit: Nvm, you're not using Aniyomi, right? If you are using MPV directly, you'll probably have to follow this guide right above your first comment:

https://github.com/bloc97/Anime4K/issues/99#issuecomment-2691034974

**Comment by akeakeakeake** [comment]
> Did you try not using input.config and leave it empty? Edit: Nvm, you're not using Aniyomi, right? If you are using MPV directly, you'll probably have to follow this guide right above your first comment:
> 
> [#99 (comment)](https://github.com/bloc97/Anime4K/issues/99#issuecomment-2691034974)

I'm not using Aniyomi, but I'm trying to get the Plex app to run MPV so I could run the shaders on my TV. First step is getting the shaders to work at all but I'm already stuck, (and i don't even know if its possible to hook up MPV with Plex on android (or Plex with Aniyomi if thats a thing)). I'll try the steps in the linked comment though!

**Comment by akeakeakeake** [comment]
Yeah unfortunately it still doesn't work :/ I'm beginning to think it doesn't work on android tv at all

**Comment by sami155** [comment]
> Yeah unfortunately it still doesn't work :/ I'm beginning to think it doesn't work on android tv at all

Have you tried the tutorial with adb I linked?

**Comment by akeakeakeake** [comment]
> > Yeah unfortunately it still doesn't work :/ I'm beginning to think it doesn't work on android tv at all
> 
> Have you tried the tutorial with adb I linked?

The comment? if so yeah. If not, i don't see anything else

**Comment by kirito201711** [comment]
> > [@rowel-eshan](https://github.com/rowel-eshan) ok, maybe its the formatting? can you please try this?
> > ```
> > # Optimized shaders for lower-end GPU: Mode B (Fast)
> > profile=gpu-hq
> > glsl-shaders="/storage/emulated/0/Documents/shaders/Anime4K_Clamp_Highlights.glsl"
> > glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Restore_CNN_Soft_M.glsl"
> > glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_M.glsl"
> > glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x2.glsl"
> > glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_AutoDownscalePre_x4.glsl"
> > glsl-shaders-append="/storage/emulated/0/Documents/shaders/Anime4K_Upscale_CNN_x2_S.glsl"
> > ```
> 
> Do you have any idea why this doesn't work for nvidia shield? Should be same path unless i'm missing something
> 
> ![Image](https://github.com/user-attachments/assets/5afbbaab-1b92-41ec-be94-4c2922dda5b1)
> 
> ![Image](https://github.com/user-attachments/assets/988b170a-ec55-4765-b0dc-742453b5ce35)

How can I check the statistics on the second page? I only have one page.

**Comment by yooooooLeon** [comment]
您的邮件已收到！谢谢！

**Comment by kirito201711** [comment]
> > > Yeah unfortunately it still doesn't work :/ I'm beginning to think it doesn't work on android tv at all
> > 
> > 
> > Have you tried the tutorial with adb I linked?
> 
> The comment? if so yeah. If not, i don't see anything else

Did you enable gpu-next? I found that enabling gpu-next causes filters on Android to not work.

**Comment by SquadCraftFTW** [comment]
If you have a Plex server, plezy client now supports anime4k and Nvidia shaders when you change to mpv backend, the implementation is really simple, just select the shader in playback settings. Hopefully somebody can use the code for a local media player.

**Comment by StepanSad** [comment]
Guys, I got this working on mpv android (needs mpv to have access to the files)

(WORKS WITH GPU-NEXT)

1. Change the left tap gesture to custom

<img width="1182" height="2560" alt="Image" src="https://github.com/user-attachments/assets/dd465033-7ba4-4341-9a37-1a562a5a97b8" />

2. Put this to mpv.conf (INSIDE THE APP SETTINGS)
```
# --- Video Output and API ---
# Using gpu-next for better shader support and performance
vo=gpu-next
gpu-api=opengl
gpu-context=android

# --- Hardware Acceleration ---
# REQUIRED: mediacodec-copy allows GLSL shaders to process the video stream
hwdec=mediacodec-copy
hwdec-codecs=all

# --- Performance and Optimization ---
# Helps avoid stutters when running heavy shaders like Anime4K
vd-lavc-threads=4
# Ensures smooth playback
dither-depth=auto

# --- General ---
save-position-on-quit=yes
```
3. Put this to input.conf (INSIDE THE APP SETTINGS)
```
# --- Anime4K: One-Button Toggle (A+A / OFF) ---
# Cycles through the shader chain and displays current status on screen.
# 0x10001 = Left Side Tap on Android.

0x10001 cycle-values glsl-shaders "/sdcard/mpv/shaders/Anime4K_Restore_CNN_M.glsl:/sdcard/mpv/shaders/Anime4K_Upscale_CNN_x2_M.glsl:/sdcard/mpv/shaders/Anime4K_Restore_CNN_S.glsl:/sdcard/mpv/shaders/Anime4K_Upscale_CNN_x2_S.glsl" ""; show-text "Anime4K" 

# --- Center Tap (0x10002) ---
0x10002 cycle pause
```
4. All shaders should be in this path: /sdcard/mpv/shaders/
Example: /sdcard/mpv/shaders/Anime4K_Upscale_CNN_x2_M.glsl

6. Open video. Double-tap the left side of the screen. Look at the picture. Now double tap again to turn it off.


(One Piece)
Anime4k on:

<img width="2560" height="1182" alt="Image" src="https://github.com/user-attachments/assets/9c5260cd-d1b4-4f89-8eee-8ad512bdd2ec" />

Anime4k off:

<img width="2560" height="1182" alt="Image" src="https://github.com/user-attachments/assets/42304180-2a71-459b-9096-ae100b596ea3" />

**Comment by yooooooLeon** [comment]
您的邮件已收到！谢谢！
